### PR TITLE
Matrix Cleanup stage 6

### PR DESCRIPTION
**Pull request by siefkenj**
Created `MatrixSubspaces` and `MatrixEigen` and `MatrixSpecial` to house relavent operations.  `MatrixSpecial` handles the creation of special matrices like `ones()`, `zeros()`, `eye()`, etc.

### CODE REVIEW

File: sympy/matrices/sparse.py [hunk 1]
```diff
@@ -403,6 +408,10 @@ def _eval_transpose(self):
     def _eval_values(self):
         return [v for k,v in self._smat.items() if not v.is_zero]
 
+    @classmethod
+    def _eval_zeros(cls, rows, cols):
```
Comment (jksuom) [inline sympy/matrices/sparse.py]: Why is this called `_eval_zeros` instead of `zeros`?

Comment (siefkenj) [inline sympy/matrices/sparse.py]: `zeros` is defined for all matrices and does error checking and has a consistent doc string.  `_eval_zeros()` does the real work and is mean to be overridden in subclasses.

Comment (jksuom) [inline sympy/matrices/sparse.py]: I didn't find it called anywhere. Can it be called only after all subclasses have it defined?

Comment (siefkenj) [inline sympy/matrices/sparse.py]: Look at `matrices.py` line 933 in this PR

Comment (jksuom) [inline sympy/matrices/sparse.py]: That looks ok (github didn't show it automatically).

File: sympy/matrices/matrices.py [hunk 2]
```diff
@@ -2648,115 +2978,484 @@ def rref(self, iszerofunc=_iszero, simplify=False, pivots=True, normalize_last=T
...
+        if primitive:
+            # if the primitive flag is set, get rid of any common
+            # integer denominators
+            def denom_clean(l):
+                from sympy import gcd
+                return [simpfunc(v / gcd(list(v))) for v in l]
+            ret = [(val, mult, denom_clean(es)) for val, mult, es in ret]
+        if has_floats:
+            # if we had floats to start with, turn the eigenvectors to floats
+            ret = [(val, mult, [v.evalf(chop=chop) for v in es]) for val, mult, es in ret]
```
Comment (jksuom) [inline sympy/matrices/matrices.py]: It seems that `val` here should also be returned as a float.

Comment (siefkenj) [inline sympy/matrices/matrices.py]: Good call!  This fixed the unfinishing test!

Comment (jksuom) [inline sympy/matrices/matrices.py]: I expected that...

**Comment by jksuom** [comment]
This looks good. Thanks.

**Pull request merged**